### DISCUSSION TITLE
Setting seeds for CutSet.from_shar, DynamicBucketingSampler, and torch.utils.data.DataLoader for multi-node training

### DISCUSSION THREAD

**Discussion by g8a9** [lhotse-speech/lhotse#1547] [Q&A]
Hi all, I'm opening this thread to ask for advice on how to set the various seeds and options for different classes while training in a multi-note, multi-gpu, multi-worker setup. I'm asking this because I notice some data duplication: in my dataset class, I write all the `cut.id` values the sampler yields to disk, and, when controlling them post-hoc, I see duplication both within and across workers.

Below, you can find code snippets to verify how I set the relevant bits:

**CutSet**

```python
cuts = CutSet.from_shar(
    in_dir=shar_path,
    shuffle_shards=True,
    seed="randomized", 
    stateful_shuffle=True,
    # split_for_dataloading=False,
)
```

**Sampler**

 ```python
sampler = DynamicBucketingSampler(
    cuts,
    constraint=constraint,
    shuffle=shuffle,  # True
    drop_last=drop_last,
    shuffle_buffer_size=shuffle_buffer_size,  # 10000
    seed="randomized",
    num_buckets=num_buckets,  # typically 30
    duration_bins=bucket_duration_bins,
    buffer_size=bucket_buffer_size,  # 10000
    rank=0,
    world_size=1,
)
```

**Dataloader**

```python
dloader_kwargs = {
    "dataset": IterableDatasetWrapper(
        dataset=dataset,
        sampler=sampler,
        # auto_increment_epoch=auto_increment_epoch,
    ),
    "worker_init_fn": make_worker_init_fn(
        rank=global_rank,
        world_size=world_size,
        seed=seed,  # 42
        set_different_node_and_worker_seeds=True,
    ),
    "persistent_workers": num_workers > 0,
}
dataloader = torch.utils.data.DataLoader(
    **dloader_kwargs,
    batch_size=None,  # Batching handled by sampler
    num_workers=num_workers,
    pin_memory=pin_memory,
    prefetch_factor=prefetch_factor if num_workers > 0 else 0,  # 2
)
```

You can consider that `world_size` and `rank` are correctly set through "WORLD_SIZE" and "RANK" env variables.
I discovered this bug because my loss increased as I used more GPUs. At the moment, I'm experimenting with a single node, two gpus, and two data workers (`num_workers=2`).

Do you have any intuition about what I'm doing wrong? I might be overlooking something extremely silly :) 
Moreover, I noticed that `CutSet.from_shar` has a `split_for_dataloading` parameter, which might be helpful, but I'm not currently using it. If needed, I can provide more info on the setup. 

Thanks!

**Comment by pzelasko** [comment]
Your code seems fine to me. With that setup, over a full iteration, there is an expected duplication of num GPUs * num workers (2 * 2 = 4 total). However, the data should be iterated in a different order on each worker+GPU combination. 

Typically we also add `cuts = cuts.repeat()` before passing that to sampler to make an infinite iterator.

You can set `split_for_dataloading` which will distribute shards between workers, but without guarantee for getting the same number of training steps - so you might want to add a `.repeat()` anyway to avoid deadlocks.

**Comment by g8a9** [comment]
Thanks for the super quick response and the checks. So, to have a full data loading pipeline that iterates _only once over the full CutSet_ where each worker, for each process (GPU) and each node, _sees different cuts_, is achieved by

1. setting `split_for_dataloading=True`
2. doing `cuts = cuts.repeat()`
3. manually computing a number of steps as `steps = batch_duration / (total_duration * world_size)` or `steps = batch_duration / (total_duration * world_size * gradient_accumulation_steps)` if the training logic supports accumulating gradients. Perhaps, this number should also be rescaled if `quadratic_duration` is set.

In that formula, `batch_duration` is the TimeConstraint `max_duration` parameter used to dynamically build batches.

Does this all make sense? 

Thanks again.

**Reply by pzelasko** [reply]
Almost - if you do `.repeat()`, now it's an infinite iterator. If you don't do `.repeat()`, you'll get a different number of batches on each GPU (the formulas for number of steps you provided are just an approximation, not the actual number - it can't be easily pre-computed). That will cause a deadlock near the end of an epoch where one GPU has no more data and the other has some batches left to process.

Your options then are:
* drop the remainder batches (requires NCCL sync to check)
* try to rebalance (again some tricks with NCCL needed)

None of them are very clean, which is why I preferred to forget about the epoch concept, just keep iterating, and set a number of steps to train.

**Reply by g8a9** [reply]
> I preferred to forget about the epoch concept, just keep iterating, and set a number of steps to train.

fully agree on this, this is why I would just approximate the epoch by setting `steps = batch_duration / (total_duration * world_size * gradient_accumulation_steps)` (sorry, the three points in the previous message were not mutually exclusive, I'd do all of them).

EDIT: probably the formula should be `steps = batch_duration / (total_duration * world_size * num_workers * gradient_accumulation_steps)`
